File: problems/042/042.f90
PROGRAM F042

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/042/p042_words.txt"
    CHARACTER (LEN=1) :: CH

    INTEGER :: COUNT_VAL =  0
    INTEGER :: INDEX_VAL = -1
    INTEGER :: WORD_VAL  =  0

    LOGICAL :: IS_INCREMENTED = .FALSE.

    REAL    :: SQRT_TERM, N_TERM

    OPEN (2, FILE = FILENAME, STATUS = 'OLD')

    READ(2, '(A1)', ADVANCE='NO', END=10) CH

    DO
        READ(2, '(A1)', ADVANCE='NO', END=10) CH

        IF ((CH >= 'A' .AND. CH <= 'Z') .OR. (CH >= 'a' .AND. CH <= 'z')) THEN

            WORD_VAL       = WORD_VAL  + IACHAR(CH) - 64
            INDEX_VAL      = INDEX_VAL + 1
            IS_INCREMENTED = .FALSE.

        ELSE

            IF (.NOT. IS_INCREMENTED) THEN

                INDEX_VAL = 0
                IS_INCREMENTED = .TRUE.

                SQRT_TERM = SQRT(1.0E0 + (8.0E0 * WORD_VAL))
                WORD_VAL = 0

                IF ((SQRT_TERM - INT(SQRT_TERM)) > 0.0) CYCLE

                N_TERM = (SQRT_TERM - 1.0) / 2.0

                IF ((N_TERM - INT(N_TERM)) > 0.0) CYCLE

                COUNT_VAL = COUNT_VAL + 1

            END IF

        END IF

    END DO

    10 CLOSE(2)

    PRINT ('(I0)'), COUNT_VAL

END PROGRAM F042

File: problems/042/042.c
#include <math.h>
#include <stdio.h>
#include <ctype.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    FILE *fp;
    char* filename = "problems/042/p042_words.txt";
    fp = fopen(filename, "r");

    char ch;
    int  count    =  0;
    int  index    = -1;
    int  word_val =  0;
    int  is_incremented = 0;

    fseek(fp, 1, SEEK_SET);
    while((ch = fgetc(fp)) != EOF)
    {
        if (isalpha(ch))
        {
            is_incremented = 0;
            word_val += ch - 64;
            index++;
        }
        else
        {
            if (!is_incremented)
            {
                index = 0;
                is_incremented = 1;

                float sqrt_term = sqrt(1 + (8 * word_val));
                word_val = 0;

                if (sqrt_term != (int)sqrt_term)
                    continue;

                float n_term = (sqrt_term - 1) / 2;

                if (n_term != (int)n_term)
                    continue;

                count++;
            }
        }
    }

    fclose(fp);

    printf("%d\n", count);

    return 0;
}
